src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlySpanND.cs

```diff
@@ -0,0 +1,468 @@
+            /// <summary>The total item count.</summary>
+            private nint _items;
+
+            /// <summary>Initialize the enumerator.</summary>
+            /// <param name="span">The span to enumerate.</param>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal Enumerator(ReadOnlySpanND<T> span)
+            {
+                _span = span;
+                _items = -1;
+                _curIndices = new nint[_span.Rank];
+                _curIndices[_span.Rank - 1] = -1;
+            }
+
+            /// <summary>Advances the enumerator to the next element of the span.</summary>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public bool MoveNext()
+            {
+                SpanNDHelpers.AdjustIndices(_span.Rank - 1, 1, ref _curIndices, _span.Shape);
+
+                if (_items < _span.Length)
+                    _items++;
+
+                return _items < _span.Length;
+            }
+
+            /// <summary>Gets the element at the current position of the enumerator.</summary>
+            public ref readonly T Current
+            {
+                [MethodImpl(MethodImplOptions.AggressiveInlining)]
+                get => ref _span[_curIndices];
+            }
+        }
+
+        /// <summary>
+        /// Returns a reference to the 0th element of the Span. If the Span is empty, returns null reference.
+        /// It can be used for pinning and is required to support the use of span within a fixed statement.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public ref readonly T GetPinnableReference()
+        {
+            // Ensure that the native code has just one forward branch that is predicted-not-taken.
+            ref T ret = ref Unsafe.NullRef<T>();
+            if (_linearLength != 0) ret = ref _reference;
+            return ref ret;
+        }
+
+        /// <summary>
+        /// Copies the contents of this read-only span into destination span. If the source
+        /// and destinations overlap, this method behaves as if the original values in
+        /// a temporary location before the destination is overwritten.
+        /// </summary>
+        /// <param name="destination">The span to copy items into.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown when the destination Span is shorter than the source Span.
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void CopyTo(SpanND<T> destination)
+        {
+            // Using "if (!TryCopyTo(...))" results in two branches: one for the length
+            // check, and one for the result of TryCopyTo. Since these checks are equivalent,
+            // we can optimize by performing the check once ourselves then calling Memmove directly.
+            if (_linearLength <= destination.Length)
+            {
+                Span<nint> curIndices = stackalloc nint[Rank];
```

Comment:
This is dangerous for large `Rank` and can stack overflow. We likely need to explicitly handle small vs large (with the expectation that almost all real world cases will be less than 5)

Reply:
@stephentoub was suggesting using an array pool in those cases. At what point do you think we should roll over for that?